compiler/rustc_resolve/src/macros.rs

```diff
@@ -500,6 +505,28 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             Err(Determinacy::Undetermined) => return Err(Indeterminate),
         };
 
+        if self.tcx.features().diagnostic_namespace
+            && kind == MacroKind::Attr
+            && !path.segments.is_empty()
+            && path.segments[0].ident.as_str() == "diagnostic"
+        {
+            if path.segments.len() != 2 {
+                self.tcx
+                    .sess
+                    .span_err(path.span, "Diagnostic attributes requires a path with 2 segments");
+                self.tcx
+                    .sess
+                    .span_err(path.span, "Diagnostic attributes requires a path with 2 segments");
+            } else {
+                self.tcx.sess.parse_sess.buffer_lint(
+                    UNKNOWN_DIAGNOSTIC_ATTRIBUTES,
+                    path.segments[1].span(),
+                    node_id,
+                    "Unknown diagnostic attribute",
+                );
+            }
+        }
```

Comment:
This is not a right place for these checks, they should probably happen somewhere in `ast_validation`.

Reply:
Can you point to a good place for this kind of check? I'm not that well informed about the details about the compiler architecture.

Reply:
`fn visit_attribute` in compiler\rustc_ast_passes\src\ast_validation.rs
It can be used for feature gating as well.